File: kit/cache-api/src/main/java/org/toolkit4j/cache/api/CacheConfig.java
/* (C)2023*/
package org.toolkit4j.cache.api;

import java.util.function.Consumer;
import java.util.function.Function;

import lombok.Builder;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import lombok.experimental.Accessors;


@Accessors(chain = true)
public interface CacheConfig<K, V> {

    default Function<K, Object> keyConvertor() {
        return (k) -> k;
    }

    default boolean isStatisticsEnabled() {
        return false;
    }

    int maxSize();

    CacheType cacheType();

    default boolean enableNullValue() {
        return false;
    }

    default Consumer<Cache<K, V>> preloader() {
        return (cache) -> {

        };
    }
}

File: cache/cache-api/src/main/kotlin/org/toolkit4j/cache/CacheConfig.kt
package org.toolkit4j.cache

import org.toolkit4j.cache.api.Cache
import org.toolkit4j.cache.api.CacheType
import java.util.function.Consumer
import java.util.function.Function

enum class CacheType {
    LOCAL,

    REMOTE
}

interface CacheConfig<K, V> {
    fun keyConvertor(): Function<K, Any?> {
        return Function<K, Any?> { k: K? -> k }
    }

    fun isStatisticsEnabled(): Boolean {
        return false
    }

    fun maxSize(): Int

    fun cacheType(): CacheType

    fun enableNullValue(): Boolean {
        return false
    }

    fun preloader(): Consumer<Cache<K?, V?>> {
        return Consumer<Cache<K?, V?>> { }
    }
}